src/wallet/coinselection.cpp

```diff
@@ -0,0 +1,159 @@
+ * of a UTXO. UTXOs are sorted by their effective values and the trees is explored deterministically per the inclusion branch first.
+ * At each node, the algorithm checks whether the selection is within the target range. While the selection has not reached the
+ * target range, more UTXOs are included. When a selection's value exceeds the target range, the complete subtree deriving from
+ * this selection can be omitted. At that point, the last included UTXO is deselected and the corresponding omission branch explored
+ * instead. The search ends after the complete tree has been searched or after a limited number of tries.
+ *
+ * The search continues to search for better solutions after one solution has been found. The best solution is chosen by minimizing
+ * the waste metric. The waste metric is defined as the cost to spend the current inputs at the given fee rate minus the long
+ * term expected cost to spend the inputs, plus the amount the selection exceeds the spending target:
+ *
+ * waste = selectionTotal - target + inputs × (currentFeeRate - longTermFeeRate)
+ *
+ * The algorithm uses two additional optimizations. A lookahead keeps track of the total value of the unexplored UTXOs. A subtree
+ * is not explored if the lookahead indicates that the target range cannot be reached. Further, it is unnecessary to test
+ * equivalent combinations. This allows us to skip testing the inclusion of UTXOs that match the effective value and waste of an
+ * omitted predecessor.
+ *
+ * The Branch and Bound algorithm is described in detail in Murch's Master Thesis: https://murch.one/wp-content/uploads/2016/11/erhardt2016coinselection.pdf
+ *
+ * @param const std::vector<CInputCoin>& utxo_pool -> The set of UTXOs that we are choosing from. These UTXOs will be sorted in descending order
+ *                                             by effective value and the CInputCoins' values are their effective values.
+ * @param const CAmount& target_value -> This is the value that we want to select. It is the lower bound of the range.
+ * @param const CAmount& cost_of_change -> This is the cost of creating and spending a change output. This plus target_value is the upper bound
+ *                                  of the range.
+ * @param std::set<CInputCoin>& out_set -> This is an output parameter for the set of CInputCoins that have been selected.
+ * @param CAmount& value_ret -> This is an output parameter for the total value of the CInputCoins that were selected.
+ * @param CAmount not_input_fees -> The fees that need to be paid for the outputs and fixed size overhead (version, locktime, marker and flag)
+ */
+
+static const size_t TOTAL_TRIES = 100000;
+
+bool SelectCoinsBnB(std::vector<CInputCoin>& utxo_pool, const CAmount& target_value, const CAmount& cost_of_change, std::set<CInputCoin>& out_set, CAmount& value_ret, CAmount not_input_fees)
+{
+    out_set.clear();
+    CAmount curr_value = 0;
+
+    int depth = 0;
```

Comment:
In commit "Implement Branch and Bound coin selection in a new file" (a406c66f1b113e6418eec3b03e0301017478634e)

I think it would make the traversal code easier to understand to get rid of `depth` variable and the trailing false entries in the `curr_selection` vector. I implemented this in cf0a82d49112e204112b54d76e661a8e66a28c8b, (fetchable with `git fetch -n https://github.com/ryanofsky/bitcoin pr/nodepth:nodepth`).

Reply:
I've used your commit. It will be squashed into a406c66